### ISSUE TITLE
Nexus interleave reading performance issue

### ISSUE THREAD

**Issue by ODiogoSilva** [ODiogoSilva/TriFusion#95]
Reading interleave nexus files takes significantly more time to run than an equivalent leave file.

**Comment by StuntsPT** [comment]
This is to be expected due to the number of lines you have to parse.
The only way to make the performance similar would be to (godforbid!) load everything into memory before parsing...

**Comment by ODiogoSilva** [comment]
Loading into memory impacts performance on the memory side and it will be impossible for large data sets. While it may not be possible to speed things up with python built-ins, I'll still try to perform the file parsing with cython or numba. Since the parsing code only deals with strings and dictionaries, I think (hope) this may be relatively simple.

Meanwhile I'll continue my quest of finding out why is interleave still a thing....

**Comment by StuntsPT** [comment]
[quote]Loading into memory impacts performance on the memory side and it will be impossible for large data sets.[/quote]
Yes, hence the "godfobid" part. This will die fast on large datasets.
Using cython code will get you into another dependency, and require compiling...
How large a performance gap are we talking about?

**Comment by ODiogoSilva** [comment]
With a data set of 110 taxa, 1.876.053 bp, (230Mb interleaved; 200Mb leave): 

- Interleave: 504s (around 8m30s)
- Leave: 3s

That's ~168x slower...

**Comment by StuntsPT** [comment]
Try my fork and see if it helps.
You're applying a lot of logic on each parsed line, my changes should reduce this.
It can still be further improved, if we dare to assume the taxa are always in the same order in the nexus file. That should make one hell of a larger difference.

**Comment by ODiogoSilva** [comment]
Thanks for the code!

However, I don't think the parsing logic should make a meaningful impact. Those checks are very fast, even for the number of lines being parsed. Somehow, the fork's code took 20 extra seconds to run, though I did not run it multiple times to see if it was significant.

I revisited the code now, which is quite old, and I've found the major culprit of the slow-down. The main dictionary is storing sequences as strings and concatenating a new string every iteration. This is quite inefficient and gets progressively worse as the size of the string increases. So, the alternative is to store each sub-sequence in a list and then use join() after the parsing is done. Doing so, reduced the run time of the previous data set to 24s, which is still significantly higher than the leave format but at least it's tolerable.
